returnn/torch/data/pipeline.py

```diff
@@ -455,6 +466,94 @@ def __getitem__(self, index):
+        *,
+        buffer_size: int,
+        monotonic_data_keys: Sequence[str],
+        seed: Optional[int] = None,
+    ):
+        """
+        :param dataset: batches dataset to shuffle
+        :param buffer_size: buffer size for shuffling
+        :param monotonic_data_keys: data keys that will be excluded from shuffling/keep their order
+        :param seed: random seed
+        """
+        super().__init__()
+
+        self._dataset = dataset
+        assert buffer_size > 0
+        self._buffer_size = buffer_size
+        self._monotonic_data_keys = monotonic_data_keys
+        self._rng = numpy.random.RandomState()
+        self._seed = seed
+
+    def __iter__(self):
+        # The implementation is very similar to the PostprocessingDataset's combinator LaplaceOrdering.
+
+        if self._seed is None:
+            # torch uses a random seed when no other seed has been set via set_seed or via __init__
+            self._seed = numpy.random.randint(0, 2**32 - 1)
+        self._rng.seed(self._seed)
+        self._seed = None
+
+        data_iter = iter(self._dataset)
+
+        batch_buffer: List[List[Dict[str, Any]]] = list(itertools.islice(data_iter, self._buffer_size))
+        has_ended = False
+
+        while True:
+            # Make sure to not reorder the monotonic values from self._monotonic_data_keys.
+            # These can contain things like complete_frac, which should be kept in order.
+            ordered_data = {
+                key: [data_dict[key] for batch in batch_buffer for data_dict in batch]
+                for key in self._monotonic_data_keys
+            }
+            self._rng.shuffle(batch_buffer)
+            for key in self._monotonic_data_keys:
+                data_dicts = [data_dict for batch in batch_buffer for data_dict in batch]
+                assert len(data_dicts) == len(ordered_data[key])
+                for ordered_value, data_dict in zip(ordered_data[key], data_dicts):
+                    data_dict[key] = ordered_value
+
+            next_batch_buffer = []
+
+            for item in batch_buffer:
+                yield item
+
+                try:
+                    if not has_ended:
+                        next_batch_buffer.append(next(data_iter))
+                except StopIteration:
+                    has_ended = True
+
+            if len(batch_buffer) < self._buffer_size:
+                assert has_ended and not next_batch_buffer
+                break
+
+            batch_buffer = next_batch_buffer
+
+    def set_seed(self, seed: int) -> ShufflingDataPipe:
+        """
+        Sets the seed for the next invocation of ``__iter__``, for compatibility with
+        ``torch.utils.data.graph_settings.apply_random_seed``.
+        """
+        self._seed = seed % (2**32)  # seed must be within [0, 2**32) for seeding RandomState
+        return self
+
+    def __getitem__(self, index):
+        raise Exception(f"{self.__class__.__name__}.__getitem__ not supported")
```

Comment:
Note: `torch.utils.data.datapipes.iter.Shuffler` also defines `__getstate__` and `__setstate__`. I'm not exactly sure why that is needed, i.e. whether/why the serialization breaks otherwise. But then the question is, if it breaks there, maybe it would also break for you? Or maybe it only breaks under certain circumstances? From a quick glance, I'm not exactly sure why it is needed. Maybe the RNG state is not properly preserved otherwise?